client/storage.c

```diff
@@ -193,6 +193,304 @@ static int sort_callback(const void *l, const void *r)
+ One thread to preload storage files.
+ The other one to add blocks to rbtree.
+ */
+#define SIMPLE_QUEUE_MAX_MEM (0x100000000) /* Max memory for queue 100MB */
+#define QUEUE_SIZE 2
+#define	QUEUE_WRITE 0x1
+#define	QUEUE_DONE 0x2
+
+struct queue_item {
+	struct xdag_block *data;
+	char filename[128];
+	size_t size;
+	struct queue_item *next;
+};
+
+struct simple_queue {
+	struct queue_item *head;
+	struct queue_item *tail;
+	size_t totalsize;
+	uint8_t status;
+};
+
+static pthread_mutex_t simple_queue_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+static struct simple_queue g_msg_queue[QUEUE_SIZE];
+
+static void simple_queue_init()
+{
+	for (int i = 0; i < QUEUE_SIZE; i++) {
+		g_msg_queue[i].head = NULL;
+		g_msg_queue[i].tail = NULL;
+		g_msg_queue[i].totalsize = 0;
+		g_msg_queue[i].status = QUEUE_WRITE;
+	}
+}
+
+static int simple_queue_push(void *data, size_t size/*, char *filename */)
+{
+	struct queue_item *item = (struct queue_item*)malloc(sizeof(struct queue_item));
+	if(!item) {
+		return -1;
+	}
+	
+	item->data = data;
+	item->size = size;
+	
+	/* strcpy(item->filename, filename); */
+	
+	item->next = NULL;
+	
+	pthread_mutex_lock(&simple_queue_mutex);
+	
+	uint8_t res = 0;
+	for (int i = 0; i < QUEUE_SIZE; i++) {
+		/* check if queue is ready for write. */
+		if(g_msg_queue[i].totalsize + size >= SIMPLE_QUEUE_MAX_MEM) {
+			g_msg_queue[i].status = !QUEUE_WRITE;
+		}
+		
+		if(g_msg_queue[i].status & QUEUE_WRITE) {
+			if(!g_msg_queue[i].head) {
+				g_msg_queue[i].head = item;
+				g_msg_queue[i].tail = item;
+			} else {
+				g_msg_queue[i].tail->next = item;
+				g_msg_queue[i].tail = item;
+			}
+			g_msg_queue[i].totalsize += item->size;
+			res = 1;
+			break;
+		}
+	}
+	
+	pthread_mutex_unlock(&simple_queue_mutex);
+	
+	if(res == 0) {
+		free(item);
+	}
+	
+	return res;
+}
+
+static struct queue_item* simple_queue_splice()
+{
+	pthread_mutex_lock(&simple_queue_mutex);
+	
+	struct queue_item * res = NULL;
+	for (int i = 0; i < QUEUE_SIZE; i++) {
+		/* check if queue is ready for write. */
+		if(g_msg_queue[i].head == NULL) {
+			g_msg_queue[i].status |= QUEUE_WRITE;
+			g_msg_queue[i].totalsize = 0;
+		} else if(!(g_msg_queue[i].status & QUEUE_WRITE)) {
+			res = g_msg_queue[i].head;
+			g_msg_queue[i].head = res->next;
+			g_msg_queue[i].totalsize -= res->size;
+			
+			if(g_msg_queue[i].tail == res) {
+				g_msg_queue[i].tail = NULL;
+				g_msg_queue[i].status |= QUEUE_WRITE;
+				g_msg_queue[i].totalsize = 0;
+			}
+			break;
+		}
+	}
+
+	pthread_mutex_unlock(&simple_queue_mutex);
+	return res;
+}
+
+struct init_storage_param {
+	xdag_time_t start_time; 
+	xdag_time_t end_time;
+	xdag_time_t *data;
+	void *(*callback)(void *, void *);
+};
+
+static void *init_storage_load_thread(void *data)
+{   
+	struct init_storage_param *param = (struct init_storage_param *)data;
+	xdag_time_t start_time = param->start_time;
+	xdag_time_t end_time = param->end_time;
+	
+	struct xdag_block readbuf[bufsize];
+	struct xdag_storage_sum s;
```

Comment:
Could you please avoid using one-char variables? IMHO "storage_sum" is more clear

Reply:
You are right, I will rename it when test done.